File: js/asyncawait.js

function async (fn) {
	return function () {
		// Get a generator by calling the function one time
		var generator = fn.apply(this, arguments);

		function callback (isFirstTime) {
			try {
				// The script has just yielded a function to await
				var [fn, args] = (isFirstTime === generator)
					? generator.next()
					: generator.send(Array.slice(arguments));

				// Run the function with the initial args + our callback
				fn.apply(this, args.concat(callback));

			} catch (error if error === StopIteration) {
				// The generator completed its execution
			}
		};

		// Use generator as a dummy value to know when
		// callback is called for the first time
		callback(generator);
	};
}

// await(fn)(1, 2, 3)
// -> [fn, [1, 2, 3]]
function await (fn) {
	return function() {
		return [ fn, Array.slice(arguments) ];
	}
}

File: src/python/asyncawait.py

# Requires Python 3

def async(fn):
	def f(*args):
		gen = fn(*args)
		def callback(*args):
			try:
				if len(args) == 1 and args[0] == callback:
					(fn, args) = next(gen)
				else:
					(fn, args) = gen.send(args)
				fn(*(args + (callback,)))
			except StopIteration:
				pass

		callback(callback)
	return f

def await(fn):
	def f(*args):
		return (fn, args)
	return f

